lls/cache.c

```diff
@@ -399,6 +399,84 @@ xmit_icmp_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
 	return 0;
 }
 
+static void
+process_icmp_ping_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
+	struct lls_config *lls_conf, struct gatekeeper_if *iface)
+{
+	uint16_t tx_queue;
+	struct token_bucket_ratelimit_state *rs;
+	unsigned int i;
+	unsigned int num_granted_pkts;
+	unsigned int num_reply = 0;
+
+	if (iface == &lls_conf->net->front) {
+		tx_queue = lls_conf->tx_queue_front;
+		rs = &lls_conf->front_icmp_rs;
+	} else {
+		tx_queue = lls_conf->tx_queue_back;
+		rs = &lls_conf->back_icmp_rs;
+	}
+
+	num_granted_pkts = tb_ratelimit_allow_n(num_pkts, rs);
+
+	for (i = 0; i < num_pkts; i++) {
+		if (num_reply >= num_granted_pkts ||
+				xmit_icmp_ping_reply(iface, pkts[i]) != 0) {
+			rte_pktmbuf_free(pkts[i]);
+			continue;
+		}
+		pkts[num_reply++] = pkts[i];
+	}
+
+	send_pkts(iface->id, tx_queue, num_reply, pkts);
+}
+
+static void
+process_icmp_pkts(struct lls_config *lls_conf, struct lls_icmp_req *icmp)
+{
+	struct rte_mbuf *ping_pkts[icmp->num_pkts];
+	unsigned int num_ping_pkts = 0;
+	int i;
+
+	for (i = 0; i < icmp->num_pkts; i++) {
+		struct rte_mbuf *pkt = icmp->pkts[i];
+		struct rte_ether_hdr *eth_hdr =
+			rte_pktmbuf_mtod(pkt, struct rte_ether_hdr *);
+		struct rte_ipv4_hdr *ip4hdr;
+		struct rte_icmp_hdr *icmphdr;
+		size_t l2_len = pkt_in_l2_hdr_len(pkt);
+
+		pkt_in_skip_l2(pkt, eth_hdr, (void **)&ip4hdr);
+		if (pkt->data_len < (ICMP_PKT_MIN_LEN(l2_len) +
+				ipv4_hdr_len(ip4hdr) - sizeof(*ip4hdr))) {
+			rte_pktmbuf_free(pkt);
+			continue;
+		}
+
+		if (rte_ipv4_frag_pkt_is_fragmented(ip4hdr)) {
+			LLS_LOG(WARNING,
+				"Received fragmented ICMP packets destined to this server at %s\n",
+				__func__);
+			rte_pktmbuf_free(pkt);
+			continue;
+		}
+
+		icmphdr = (struct rte_icmp_hdr *)ipv4_skip_exthdr(ip4hdr);
+
+		if (icmphdr->icmp_type == ICMP_ECHO_REQUEST_TYPE &&
+				icmphdr->icmp_code == ICMP_ECHO_REQUEST_CODE) {
+			ping_pkts[num_ping_pkts++] = pkt;
+		} else {
+			LLS_LOG(INFO, "Received ICMP packet with type %hhu and code %hhu on %s interface\n",
+				icmphdr->icmp_type, icmphdr->icmp_code, icmp->iface->name);
+		}
```

Comment:
This pull request was motivated to report a wrong MTU on the path from Gatekeeper to Grantor servers. So add an `else if () {` to properly report that case before the general `} else {` case.